### PR TITLE
Added helper functions for handling The Siren's Steal attack

### PR DESCRIPTION
**Pull request by KyojoH**

### CODE REVIEW

File: src/features/sirenHelpers.ts [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+    );
+  };
+  return (
+    sirenHelperData.persistent.familiarBlacklist.find(compareBlacklistEntry) !==
+    undefined
+  );
+}
+
+/**
+ *
```
Comment (Zamiell) [inline src/features/sirenHelpers.ts:109]: bug

File: src/features/sirenHelpers.ts [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+ * @param familiar The familiar to check
+ * @returns Returns whether the familiar has been stolen by The Siren
+ */
+export function isStolen(familiar: EntityFamiliar): boolean {
```
Comment (Zamiell) [inline src/features/sirenHelpers.ts:113]: how about hasSirenStolen

File: src/features/sirenHelpers.ts [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+      GetPtrHash(helper.Target) === GetPtrHash(familiar)
+    ) {
+      return helper;
+    }
+  }
+  return undefined;
+}
+
+/** @hidden */
+export function init(incomingMod: Mod): void {
```
Comment (Zamiell) [inline src/features/sirenHelpers.ts:67]: put init at the top under the v declaration

File: src/features/sirenHelpers.ts [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+  }
+  sirenHelperData.persistent.familiarBlacklist.push([
+    familiarVariant,
+    familiarSubType !== undefined ? familiarSubType : -1,
+  ]);
+}
+
+/**
+ * @param familiar The familiar to check.
+ * @returns Returns the hidden 'Siren Helper' entity corresponding to the given familiar, if it
```
Comment (Zamiell) [inline src/features/sirenHelpers.ts:48]: use complete sentances

File: src/features/sirenHelpers.ts [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+  persistent: { familiarBlacklist: Array<[number, number]> };
+} = {
+  persistent: { familiarBlacklist: [] },
+};
+
+/**
+ * Blacklists a familiar from being stolen by The Siren. For example, if your character starts with
+ * one and needs it to play.
+ *
+ * @param familiarVariant The familiar variant to blacklist
```
Comment (Zamiell) [inline src/features/sirenHelpers.ts:16]: same here

File: src/features/sirenHelpers.ts (hunk 1 above)
Comment (KyojoH) [reply src/features/sirenHelpers.ts:111]: what's the bug? i don't see one

File: src/features/sirenHelpers.ts [hunk 6]
```diff
@@ -0,0 +1,116 @@
+import { getUpgradeErrorMsg } from "../errors";
+import { saveDataManager } from "./saveDataManager/main";
+
+const FEATURE_NAME = "siren helpers";
+let mod: Mod | null = null;
+const sirenHelperData: {
```
[the remark this answers is not in the document]
Comment (KyojoH) [reply src/features/sirenHelpers.ts:7]: it's an int, although [int,int][] gets fixed by eslint or prettier to Array<[int,int]>

File: src/features/sirenHelpers.ts [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+    );
+  };
+  return (
+    sirenHelperData.persistent.familiarBlacklist.find(compareBlacklistEntry) !==
+    undefined
+  );
+}
+
+/**
+ * This function checks whether the Siren has stolen a familiar.
```
Comment (KyojoH) [inline src/features/sirenHelpers.ts:108]: yes

File: src/features/sirenHelpers.ts [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+ *
+ * @param familiar The familiar to be checked.
+ * @returns Returns whether the familiar has been stolen by The Siren.
+ */
+export function hasSirenStolen(familiar: EntityFamiliar): boolean {
```
Comment (Zamiell) [inline src/features/sirenHelpers.ts]: hasSirenStolenFamiliar

File: src/features/sirenHelpers.ts [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+    familiar.Variant,
+    familiar.SubType,
+  ];
+  if (!blacklistEntryExists(newBlacklistEntry)) {
+    getSirenHelper(familiar)?.Remove();
+    familiar.AddToFollowers();
+  }
+}
+
+function blacklistEntryExists(newBlacklistEntry: [number, number]) {
```
Comment (Zamiell) [inline src/features/sirenHelpers.ts:94]: use a named tuple + object destructuring. this is a no magic zone

File: src/features/sirenHelpers.ts [hunk 10]
```diff
@@ -0,0 +1,115 @@
+import { getUpgradeErrorMsg } from "../errors";
+import { saveDataManager } from "./saveDataManager/main";
+
+const FEATURE_NAME = "sirenHelpers";
+
+let mod: Mod | null = null;
+const sirenHelperData = {
```
Comment (Zamiell) [inline src/features/sirenHelpers.ts:7]: use the const v convention. v is short for local variables and is the same in every file

**Review by Zamiell — CHANGES_REQUESTED** [review]

File: src/features/sirenHelpers.ts (hunk 10 above)
Comment (KyojoH) [reply src/features/sirenHelpers.ts:7]: done

File: src/features/sirenHelpers.ts (hunk 9 above)
Comment (KyojoH) [reply src/features/sirenHelpers.ts:94]: done

File: src/features/sirenHelpers.ts (hunk 8 above)
Comment (KyojoH) [reply src/features/sirenHelpers.ts]: done

File: src/features/sirenHelpers.ts [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+    familiarSubType !== undefined ? familiarSubType : -1,
+  ]);
+}
+
+/**
+ * @param familiar The familiar to be checked.
+ * @returns Returns the hidden 'Siren Helper' entity corresponding to the given familiar, if it
+ * exists
+ */
+export function getSirenHelper(familiar: EntityFamiliar): Entity | undefined {
```
[the remark this answers is not in the document]
Comment (KyojoH) [reply src/features/sirenHelpers.ts:62]: done

File: src/features/sirenHelpers.ts [hunk 12]
```diff
@@ -0,0 +1,116 @@
+import { getUpgradeErrorMsg } from "../errors";
+import { saveDataManager } from "./saveDataManager/main";
+
+const FEATURE_NAME = "siren helpers";
+let mod: Mod | null = null;
```
[the remark this answers is not in the document]
Comment (KyojoH) [reply src/features/sirenHelpers.ts:6]: done

File: src/features/sirenHelpers.ts [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+
+/**
+ * Blacklists a familiar from being stolen by The Siren. For example, if your character starts with
+ * one and needs it to play.
+ *
+ * @param familiarVariant The familiar variant to blacklist
+ * @param familiarSubType The subtype to blacklist. default is to blacklist all subtypes of the
+ * given variant.
+ */
+export function forceNoSiren(
```
[the remark this answers is not in the document]
Comment (KyojoH) [reply src/features/sirenHelpers.ts:31]: changed

**Pull request merged**